Add CppInterOp API dispatch mechanism

aaronj0 opened the pull request:
This PR defines the mechanism which enables dispatching CppInterOp's API without linking to it, preventing any LLVM or Clang symbols from being leaked into the client application.

This allows us to run cppyy without linking to CppInterOp motivated by the use case in ROOT. Can be used to deploy CppInterOp in an environment where dynamic linking is not favourable and the only option is dlopen'ing `libClangCppInterOp.so` with `RTLD_LOCAL`

aaronj0 committed:
Add CppInterOp API dispatch mechanism

This defines the mechanism which enables dispatching of the CppInterOp API without linking to it, preventing any LLVM or Clang symbols from being leaked into the client application. Can be used to deploy CppInterOp in an environment where dynamic linking is not favourable and the only option is dlopen'ing `libClangCppInterOp.so` with `RTLD_LOCAL`

mcbarton commented on unittests/CMakeLists.txt:
```diff
@@ -34,24 +34,24 @@ function(add_cppinterop_unittest name)
-else()
-  add_test(NAME cppinterop-${name} COMMAND ${name})
-endif()
+  if(WIN32)
```
You have have an entire commit for this format change to cmake, but I have no idea what it has to do with adding a CppInterOp API for a  dispatch mechanism. Please revert this change and put it a separate PR if you think this change needs to be made

aaronj0 replied:
As this is a draft, I have included commits that will move into a seperate PR

vgvassilev commented on include/CppInterOp/CppInterOpDispatch.h:
```diff
@@ -0,0 +1,520 @@
+// a consistent way. This is used as our dispatched API list, along with the
+// name-address pair table
+#define FOR_EACH_CPP_FUNCTION(DO)                                              \
+  DO(CreateInterpreter)                                                        \
```
This can take a pair <`CreateInterpreter`, `Type`> and in this case you can write the `using` clauses automatically. We can also use clang-tablegen to generate these for every function in the namespace `Cpp`.

aaronj0 replied:
Addressed

aaronj0 committed:
Use X macros to generate funcptrs and symbol-address dispatch table

Rename CppInterOpDispatch -> CppDispatch, Address more review comments and drop `DECLARE_CPP` style macros

vgvassilev commented on include/CppInterOp/CppInterOpDispatch.h:
```diff
@@ -0,0 +1,597 @@
+// Configured by CMake, can be overridden by defining before including this
+// header
+#ifndef CPPINTEROP_LIBRARY_PATH
+#define CPPINTEROP_LIBRARY_PATH "@CPPINTEROP_LIBRARY_PATH@"
```
That is still a problem, however, I do not understand why you need to define it if it is provided from the outside...

aaronj0 replied:
Dropped

vgvassilev commented on include/CppInterOp/CppInterOpDispatch.h:
```diff
@@ -0,0 +1,298 @@
+      return;
+    }
+
+    getCppProcAddress = reinterpret_cast<void* (*)(const char*)>(
```
```suggestion
    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
    getCppProcAddress = reinterpret_cast<void* (*)(const char*)>(
```

vgvassilev commented on include/CppInterOp/CppInterOpDispatch.h:
```diff
@@ -0,0 +1,298 @@
+/// (RTLD_LOCAL) \param[in] customLibPath Optional custom path to
+/// libclangCppInterOp.so \returns true if initialization succeeded, false
+/// otherwise
+inline bool init_functions(const char* customLibPath = nullptr) {
```
That should be a test case, right?

vgvassilev commented on unittests/CppInterOp/CMakeLists.txt:
```diff
@@ -11,6 +11,14 @@ else()
+# Add the DispatchAPITest only when building shared libraries
+if (BUILD_SHARED_LIBS)
+  set_source_files_properties(DispatchAPITest.cpp PROPERTIES COMPILE_DEFINITIONS
+    "CPPINTEROP_LIB_DIR=\"${CMAKE_BINARY_DIR}/lib/libclangCppInterOp${CMAKE_SHARED_LIBRARY_SUFFIX}\""
```
Please write a test and try to run that test against installed CppInterOp with the build folder deleted.

vgvassilev commented on unittests/CppInterOp/DispatchAPITest.cpp:
```diff
@@ -0,0 +1,57 @@
+
+#include "gtest/gtest.h"
+
+#include <dlfcn.h>
```
You probably do not need this header file.

aaronj0 replied:
> You probably do not need this header file.

Thanks, that's a remnant from earlier

aaronj0 commented on lib/CppInterOp/CMakeLists.txt:
```diff
@@ -109,6 +109,7 @@ endif()
   add_llvm_library(clangCppInterOp
     DISABLE_LLVM_LINK_LLVM_DYLIB
     CppInterOp.cpp
+    CppInterOpDispatch.cpp
```
`#include <CppInterOp/DyldDispatch.h`?

aaronj0 committed:
Ensure `Cpp` namespace is consistent between CppInterOp.h and CppDispatch.h

With the introduction of the CppStatic namespace we conditionally alias `Cpp` to the implementation provided by CppInterOp.h (link to clangCppInterOp.so) or CppDispatch (dlopen the dylib with RTLD_LOCAL)
This is an intentional ODR violation but remains consistent with our API model.

aaronj0 committed:
Rename `CppDispatch` -> `Dispatch` and `CppStatic` -> `CppImpl`

aaronj0 committed:
Improve tests, rename loader API, add unload API

vgvassilev commented on include/CppInterOp/Dispatch.h:
```diff
@@ -0,0 +1,252 @@
+//--------------------------------------------------------------------*- C++ -*-
+// CppInterOp Dispatch Mechanism
+// author:  Aaron Jomy <aaron.jomy@cern.ch>
```
```suggestion
// CppInterOp Dispatch Mechanism
```

vgvassilev commented on include/CppInterOp/Dispatch.h:
```diff
@@ -0,0 +1,267 @@
+      return;
+    }
+
+    getCppProcAddress = reinterpret_cast<void* (*)(const char*)>(
```
Add a nolint clause here.

aaronj0 replied:
Done

aaronj0 committed:
Review comments, code optimizations and other improvements

aaronj0 committed:
Make unittests Utils configureable for API Dispatch mode

aaronj0 committed:
Add unittest executable that runs entirely in Dispatch mode

Gives the mechanism complete coverage. Now runs CppInterOpTests and CppInterOpDispatchTests

aaronj0 committed:
Add default arguments, use dynamic names in the test fixtures

This patch also improves the readability of the test run log:

```
2: [ RUN      ] CppInterOpDispatchTest/InProcessJIT.Interpreter_IncludePaths
2: [       OK ] CppInterOpDispatchTest/InProcessJIT.Interpreter_IncludePaths (0 ms)
2: [ RUN      ] CppInterOpDispatchTest/InProcessJIT.Interpreter_CodeCompletion
2: [       OK ] CppInterOpDispatchTest/InProcessJIT.Interpreter_CodeCompletion (86 ms)
2: [ RUN      ] CppInterOpDispatchTest/InProcessJIT.Interpreter_ExternalInterpreter
2: [       OK ] CppInterOpDispatchTest/InProcessJIT.Interpreter_ExternalInterpreter (1123 ms)
2: [ RUN      ] CppInterOpDispatchTest/InProcessJIT.Jit_InsertOrReplaceJitSymbol
2: [       OK ] CppInterOpDispatchTest/InProcessJIT.Jit_InsertOrReplaceJitSymbol (56 ms)
2: [ RUN      ] CppInterOpDispatchTest/InProcessJIT.Jit_StreamRedirect
2: [       OK ] CppInterOpDispatchTest/InProcessJIT.Jit_StreamRedirect (0 ms)
2: [ RUN      ] CppInterOpDispatchTest/InProcessJIT.Jit_StreamRedirectJIT
2: [       OK ] CppInterOpDispatchTest/InProcessJIT.Jit_StreamRedirectJIT (154 ms)
2: [ RUN      ] CppInterOpDispatchTest/InProcessJIT.ScopeReflection_IsEnumScope
2: [       OK ] CppInterOpDispatchTest/InProcessJIT.ScopeReflection_IsEnumScope (21 ms)
2: [ RUN      ] CppInterOpDispatchTest/InProcessJIT.ScopeReflection_IsEnumConstant
```

instead of:

```
2: [ RUN      ] CppInterOpDispatchTest/InProcessJIT.InterpreterTestIncludePaths
2: [       OK ] CppInterOpDispatchTest/InProcessJIT.InterpreterTestIncludePaths (0 ms)
2: [ RUN      ] CppInterOpDispatchTest/InProcessJIT.InterpreterTestCodeCompletion
2: [       OK ] CppInterOpDispatchTest/InProcessJIT.InterpreterTestCodeCompletion (86 ms)
2: [ RUN      ] CppInterOpDispatchTest/InProcessJIT.InterpreterTestExternalInterpreter
2: [       OK ] CppInterOpDispatchTest/InProcessJIT.InterpreterTestExternalInterpreter (1123 ms)
2: [ RUN      ] CppInterOpDispatchTest/InProcessJIT.JitTestInsertOrReplaceJitSymbol
2: [       OK ] CppInterOpDispatchTest/InProcessJIT.JitTestInsertOrReplaceJitSymbol (56 ms)
2: [ RUN      ] CppInterOpDispatchTest/InProcessJIT.JitTestStreamRedirect
2: [       OK ] CppInterOpDispatchTest/InProcessJIT.JitTestStreamRedirect (0 ms)
2: [ RUN      ] CppInterOpDispatchTest/InProcessJIT.JitTestStreamRedirectJIT
2: [       OK ] CppInterOpDispatchTest/InProcessJIT.JitTestStreamRedirectJIT (154 ms)
2: [ RUN      ] CppInterOpDispatchTest/InProcessJIT.ScopeReflectionTestIsEnumScope
2: [       OK ] CppInterOpDispatchTest/InProcessJIT.ScopeReflectionTestIsEnumScope (21 ms)
2: [ RUN      ] CppInterOpDispatchTest/InProcessJIT.ScopeReflectionTestIsEnumConstant
```

We repeated the word "Test" twice and the lack of a spacer made it hard to read. Since this patch changes the TYPED_TEST line to use `CPPINTEROP_TEST_MODE` instead of `CppInterOpTest` this is a good opportunity to improve this without bloating the history

aaronj0 committed:
Address various review comments

Simplify the unittest cmake logic on top of #770
clang-format, clang-tidy suggestions

aaronj0 committed:
Try using Windows ABI for dlopen/dlsym/dlclose

aaronj0 committed:
Try undef LoadLibrary macro to prevent conflicats on windows

aaronj0 committed:
more clang-tidy

aaronj0 commented on include/CppInterOp/CppDispatch.h:
```diff
@@ -0,0 +1,597 @@
+#include <vector>
+
+#include <cstdlib>
+#include <dlfcn.h>
```
The windows specific code has been added and compiled but is untested due to the absence of a shared lib build on the CI (or so it seems) The CppInterOpDispatch test executable is not built.

aaronj0 commented:
Not sure how dynamic loading works on emscripten but the following error indicates something wrong with that step for the dispatch test executable:

```
[ 95%] Linking CXX executable CppInterOpTestsDispatch.html
wasm-ld: error: attempted static link of dynamic object ../../lib/libclangCppInterOp.so
em++: error: '/Users/runner/work/CppInterOp/CppInterOp/emsdk/upstream/bin/wasm-ld -o CppInterOpTestsDispatch.wasm CMakeFiles/CppInterOpTestsDispatch.dir/EnumReflectionTest.cpp.o CMakeFiles/CppInterOpTestsDispatch.dir/FunctionReflectionTest.cpp.o CMakeFiles/CppInterOpTestsDispatch.dir/InterpreterTest.cpp.o CMakeFiles/CppInterOpTestsDispatch.dir/JitTest.cpp.o CMakeFiles/CppInterOpTestsDispatch.dir/ScopeReflectionTest.cpp.o CMakeFiles/CppInterOpTestsDispatch.dir/TypeReflectionTest.cpp.o CMakeFiles/CppInterOpTestsDispatch.dir/Utils.cpp.o CMakeFiles/CppInterOpTestsDispatch.dir/VariableReflectionTest.cpp.o CMakeFiles/CppInterOpTestsDispatch.dir/main.cpp.o ../googletest-prefix/src/googletest-build/lib//libgtest.a ../googletest-prefix/src/googletest-build/lib//libgmock.a ../../lib/libclangCppInterOp.so -L/Users/runner/work/CppInterOp/CppInterOp/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten -L/Users/runner/work/CppInterOp/CppInterOp/emsdk/upstream/emscripten/src/lib -lGL-getprocaddr -lal -lhtml5 -lstubs -lnoexit -lc -ldlmalloc -lcompiler_rt -lc++-noexcept -lc++abi-noexcept -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr /var/folders/sz/cwqy23057413nkkfyjbx066h0000gn/T/tmp68oq2eqelibemscripten_js_symbols.so --strip-debug --export=_emscripten_stack_alloc --export=__wasm_call_ctors --export=emscripten_stack_get_current --export=_emscripten_stack_restore --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export-if-defined=main --export-if-defined=__main_argc_argv --export-table -z stack-size=65536 --no-growable-memory --initial-heap=16777216 --no-entry --table-base=1 --global-base=1024' failed (returned 1)
make[3]: *** [unittests/CppInterOp/CppInterOpTestsDispatch.html] Error 1
make[2]: *** [unittests/CppInterOp/CMakeFiles/CppInterOpTestsDispatch.dir/all] Error 2
```
